README
tableau-confluence-plugin readme file:

topics:
* Tableau Server Configuration
* Installation
* Configuration
* Embed tableau reports in confluence

Tableau Server Configuration
 - Alter httpd template
	By default, tableau runs either in SSPI mode or you can configure trusted authentication for authentication. But what if you want to do both?
	In our situation we use a plugin to export reports on confluence pages, so confluence needs to authenticate with the user credentials of the currently logged on user. And normal users should be able to use SSPI.
	  
	To enable both features alter the file with filename: httpd.conf.templ
	  
	Tableau 6.x
		1. Open the template file: 
			C:\Program Files (x86)\Tableau Server\6.0\templates
		2. Search for entry: 
			<Location ~ "^/(?!manual|auth).*">
		3. Add 'trusted' to the regular expression: 
			<Location ~ "^/(?!manual|auth|trusted).*">
		
	Tableau 7.x 
		1. Open the template file: 
			C:\Program Files (x86)\Tableau Server\7.0\templates
		2. Search for the entry: 
			<Location ~ "^/(?!(t/[^/]*/)?(manual|auth|dataserver)).*">
		3. Add 'trusted' and 'vizql/trusted' to the regular expression: 
			<Location ~ "^/(?!(t/[^/]*/)?(manual|auth|trusted|vizql/trusted|dataserver)).*">

 - Disable client ip checking
		Run following tabadmin commands:
		Note tabadmin can be found here: "C:\Program Files (x86)\Tableau\Tableau Server\7.0\bin"
		
			tabadmin set wg_server.extended_trusted_ip_checking false
			tabadmin configure
			tabadmin restart
		
 - Whitelist your confluence server(s)
		Run following tabadmin commands:
			tabadmin set wgtrusted_hosts "192.168.1.1, 192.168.1.2" (where 192.168.1.x are your confluence hosts)
			tabadmin configure
			tabadmin restart
		
		The comma separated list of ip addresses should be within double quotes with a space after each comma.
		
		For more information see the tableau server admin reference (page 224 / 225 / 229):
		http://downloads.tableausoftware.com/quickstart/server-guides/en-us/server_admin7.0.pdf
 
Installation of plugin
 - Get the jar
	- Download the latest jar file from: https://plugins.atlassian.com/plugins/tableau-plugin
	- Or build it from the source : https://github.com/schubergphilis/tableau-confluence-plugin
 - Login as a confluence admin
 - Goto the menu 'Browse'
 - Choose 'Confluence Admin'
 - In the section 'Configuration' choose 'Plugins'
 - Click on the 'Install' tab
 - And choose 'Upload plugin'
 - After this process you should see within 'Manage Existing' the 'Schuberg Philis Tableau Plugin'
 
Configuration of plugin
 Now that you have installed the plugin, configure your tableau hosts
 
 Use a key, for instance: 'prod' or 'production' and a value 'http://tableauserver' or 'https://tableauserver'
 It's possible to enter multiple hosts, but the first entry is the default entry.
 
 There is one special entry for debugging purposes, and that is the key 'debugusername', with this entry you can override the username that is used for trusted authentication.
 
 If your tableau instance requires a domain to authenticate you users, you can configure it with the key 'domain', for instance 'mydomain.local'.
 
Embed tableau reports in confluence

	Possible parameters:
	
	Parameter	Default value	Description
	title 						name of report, this title is shown when opening the full page interactive view from the 'interactive button'
	workbook 	  				workbook name on the Tableau server (case sensitive)
	report 	  					report name in the workbook (case sensitive)
	height 		550 			height in number of pixels
	width 		1280 			width in number of pixels
	environment prod 			option to point to other environment configured in section 'Configuration of plugin', for example use acc for acceptance environment
	interactive false 			opens sheet in interactive mode 
	embed 		true 			embed in page 
	toolbar 	true 			show or hide the toolbar
	tabs 		false 			show tabs for all reports within the workbook
	borderstyle 	  			extra css borderstyle properties, like 'border:red 1px solid' or 'background-color:gray'
	button 		false 			option to add a button 'Interactive view' to the non interactive (png) view and open the interactive view in a window
	noprint 	false 			when setting this option to true, this report will be skipped when printing / exporting to pdf or word 
	parameters					send extra parameters to the report
	refresh		false			refresh will invalidate the tableau cached png 
	
	Basically there are two modes
		* image view, a plain png type of image
		* interactive view, where end users can use parameters, filters, etc provided within the sheet
	
	Examples:
	
	Basic example:
	{tableau-plugin:workbook=TestWorkbook|report=Sheet1}
	
	Set width and height
	{tableau-plugin:workbook=TestWorkbook|report=Sheet1|width=1500|height=800}
	
	Open interactive view and point to acceptance environment
	{tableau-plugin:workbook=TestWorkbook|report=Sheet1|interactive=true|environment=acc}
	
	Example of using parameters	
	{tableau-plugin:workbook=TestWorkbook|report=Sheet1|parameters=colour=blue&othervalue=100}

CHANGELOG.TXT
############################## 29/12/2013 v0.92 ##############################
- migration from wiki to xhtml based storage format code added

############################## 29/12/2013 v0.91 ##############################
- renamed parameter report to view

############################## 06/12/2013 v0.9 ###############################
- added the option to disable trusted authentication (only client side)
  use the key 'disableclienttrustedauth' : 'true' to use this feature
- added the option to define an extra internal url for requesting a ticket for
  authentication, for example if your tableau server is defined as
  'prod': 'http://tableauserver' you can use the environment key (prod) by
  adding -internal, for example 'prod-internal' : 'http://tableauinternal'
- bugfix: the button show interactive view did not work
- tested with confluence 5.4

############################## 29/01/2013 v0.8 ###############################
- Multi-tenant (multi-site) support

############################## 06/04/2012 v0.7 ###############################
- added the option to refresh a view ( this will invalidate the cached png )
- added the tableau icon
- added option to add a domain to the username
- extra unit tests for the macro

############################## 03/06/2012 v0.63 ###############################
- bugfix: ampersant missing when using parameters in interactive mode
- moved code to github

############################## 10/24/2011 v0.62 ###############################
- Don't show errors in preview mode
- Optional parameters parameter

############################## 10/18/2011 v0.61 ###############################
- bugfix: authentication issue when user is not authenticated against the
  tableau server, png will load with 401 error, now trusted authentication
  is used for displaying png, interactive and export context.

############################## 09/27/2011 v0.6 ###############################
- bugfix: authentication issue when user is not authenticated against the
  tableau server, png will load with 401 error

############################## 08/22/2011 v0.5 ###############################
- Trusted authentication for export context

############################## 06/22/2011 v0.4 ###############################
- removed border
- option to add border style
- names of parameters workbook and report can contain spaces
- button is aligned left
- tabs parameter added
- if embed parameter is false, interactive view will be opened in a new window
- unit tests added
- license added
- changelog added

############################## 06/19/2011 v0.03 ##############################
- Initial version


src/main/java/com/schubergphilis/confluence/plugins/TableauMacro.java
/**
 * Developer: Roel Gerrits
 * Company: Schuberg Philis
 * Date: 3/30/11
 * Time: 10:49 AM
 *
 *      Licensed to the Apache Software Foundation (ASF) under one
 *      or more contributor license agreements.  See the NOTICE file
 *      distributed with this work for additional information
 *      regarding copyright ownership.  The ASF licenses this file
 *      to you under the Apache License, Version 2.0 (the
 *      "License"); you may not use this file except in compliance
 *      with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *      Unless required by applicable law or agreed to in writing,
 *      software distributed under the License is distributed on an
 *      "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *      KIND, either express or implied.  See the License for the
 *      specific language governing permissions and limitations
 *      under the License.
 **/

package com.schubergphilis.confluence.plugins;

import com.atlassian.bandana.BandanaManager;
import com.atlassian.confluence.user.AuthenticatedUserThreadLocal;
import com.atlassian.plugin.webresource.WebResourceManager;
import com.atlassian.renderer.RenderContext;
import com.atlassian.renderer.v2.RenderMode;
import com.schubergphilis.confluence.configuration.DefaultValueBehaviour;
import com.schubergphilis.confluence.configuration.ConfigurationManager;
import com.schubergphilis.confluence.exceptions.AuthenticationException;
import com.schubergphilis.confluence.exceptions.ValidationException;
import com.schubergphilis.confluence.html.TableauRenderer;
import com.schubergphilis.tableau.authentication.TrustedAuthentication;
import com.schubergphilis.util.http.HttpRequest;

import java.io.IOException;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.util.Map;

public class TableauMacro extends SbpBaseMacro
{
    public TableauMacro(WebResourceManager webResourceManager, BandanaManager bandanaManager)
    {
        super._webResourceManager = webResourceManager;
        super._bandanaManager = bandanaManager;
    }

    public boolean hasBody()
    {
        return false;
    }

    public RenderMode getBodyRenderMode()
    {
        return RenderMode.NO_RENDER;
    }

    public String renderPlugin(Map params, String outputType)
            throws ValidationException, AuthenticationException, IOException, NoSuchAlgorithmException, KeyManagementException
    {
        String workbook = getStrParameter(params, "workbook");
        String report = getStrParameter(params, "report");
        String view = getStrParameter(params, "view", report);
        String title = getStrParameter(params, "title", view);
        String environment = getStrParameter(params, "environment", "prod");
        String borderStyle = getStrParameter(params, "borderstyle", "");
        Integer height = getIntParameter(params, "height", 550);
        Integer width = getIntParameter(params, "width", 1280);
        Boolean interactive = getBoolParameter(params, "interactive", false);
        Boolean toolbar = getBoolParameter(params, "toolbar", true);
        Boolean embed = getBoolParameter(params, "embed", true);
        Boolean tabs = getBoolParameter(params, "tabs", false);
        Boolean button = getBoolParameter(params, "button", false);
        Boolean noPrint = getBoolParameter(params, "noprint", false);
        String parameters = getStrParameter(params, "parameters", "");
        Boolean refresh = getBoolParameter(params, "refresh", false);
        String site = getStrParameter(params, "site","");

        String externalHost = getConfigurationManager().getValue(environment, DefaultValueBehaviour.firstInList);
        String internalHost = getConfigurationManager().getValue(environment + "-internal");
        String disableClientTrustedAuthValue = getConfigurationManager().getValue("disableclienttrustedauth", "false");
        boolean disableClientTrustedAuth = disableClientTrustedAuthValue.toLowerCase().startsWith("t");

        Boolean isExportContext = isPdfOrWordOutput(outputType);

        // skip for exporting to pdf/word or preview mode
        if(isExportContext && noPrint)
            return "";

        if(RenderContext.PREVIEW.equals(outputType) && ( workbook.length() == 0 || view.length() == 0))
            return "Please provide a workbook and a report and hit the refresh button.";

        TableauRenderer renderer = getTableauRenderer()
            .withSize(width, height)
            .withView(view)
            .withWorkbook(workbook)
            .withTitle(title)
            .withInteractiveStart(interactive)
            .withEmbed(embed)
            .withToolbar(toolbar)
            .withBorderStyle(borderStyle)
            .withInteractiveButton(button)
            .withExportContext(isExportContext)
            .withTabs(tabs)
            .withRefresh(refresh)
            .withSite(site)
            .withParameters(parameters);

        determineHost(externalHost, internalHost, isExportContext, disableClientTrustedAuth, site, renderer);

        return renderer.render();
    }

    private String getHostForBackendCommunication(String internalHost, String externalHost) {
        if(internalHost != null && internalHost.length() > 0){
             return internalHost;
        }

        return externalHost;
    }

    private String getHostForRenderer(String internalHost, String externalHost, boolean isExportContext) {
        // export should use internal when configured
        if(isExportContext) {
            return getHostForBackendCommunication(internalHost, externalHost);
        }
        return externalHost;
    }

    private void determineHost(String externalHost, String internalHost, boolean isExportContext, boolean disableClientTrustedAuth, String site, TableauRenderer renderer)
            throws IOException, AuthenticationException, NoSuchAlgorithmException, KeyManagementException, ValidationException
    {
        String domain = getConfigurationManager().getValue("domain");
        String username = getUsername();

        if(domain != null && domain.length() > 0)
        {
            username = String.format("%s\\%s", domain, username);
        }

        // for debugging purposes, override the confluence username
        String debugUsername = getConfigurationManager().getValue("debugusername");
        if(debugUsername != null && debugUsername.length() > 0)
            username = debugUsername;

        // only use external host for authentication when no internal is set
        String hostForTrustedAuthentication = getHostForBackendCommunication(internalHost, externalHost);

        String ticket = null;

        if(!disableClientTrustedAuth || isExportContext) {
            ticket = getTrustedAuthentication()
                    .withTableauUrl(hostForTrustedAuthentication)
                    .withUsername(username)
                    .withSite(site)
                    .authenticate();
        }

        String hostForRenderer = getHostForRenderer(internalHost, externalHost, isExportContext);
        renderer.withHost(hostForRenderer, ticket);
    }

    private boolean isPdfOrWordOutput(String outputType)
    {
        return RenderContext.PDF.equals(outputType) || RenderContext.WORD.equals(outputType);
    }

    @Override
    void includeResources()
    {
        String resource = "tableau-plugin:javascript-resources";
        _webResourceManager.requireResource(resource);
    }

    private TableauRenderer _tableauRenderer;
    private TableauRenderer getTableauRenderer()
    {
        if(_tableauRenderer == null)
            return new TableauRenderer();

        return _tableauRenderer;
    }

    public void setTableauRenderer(TableauRenderer tableauRenderer)
    {
        _tableauRenderer = tableauRenderer;
    }

    private TrustedAuthentication _trustedAuthentication;
    private TrustedAuthentication getTrustedAuthentication()
    {
        if(_trustedAuthentication == null)
            return new TrustedAuthentication(new HttpRequest());

        return _trustedAuthentication;
    }

    public void setTrustedAuthentication(TrustedAuthentication authentication)
    {
        _trustedAuthentication = authentication;
    }

    private ConfigurationManager _configurationManager;
    private ConfigurationManager getConfigurationManager()
    {
        if(_configurationManager == null)
            return new ConfigurationManager(_bandanaManager);

        return _configurationManager;
    }

    public void setConfigurationManager(ConfigurationManager configurationManager)
    {
        _configurationManager = configurationManager;
    }

    private String _username;
    private String getUsername()
    {
        if(_username == null)
            return AuthenticatedUserThreadLocal.getUsername();

        return _username;
    }

    public void setUsername(String username)
    {
        _username = username;
    }
}